libraries/AP_HAL/CANIface.cpp

```diff
@@ -70,28 +77,59 @@ int16_t AP_HAL::CANIface::receive(CANFrame& out_frame, uint64_t& out_ts_monotoni
  */
 int16_t AP_HAL::CANIface::send(const CANFrame& frame, uint64_t tx_deadline, CanIOFlags flags)
 {
-    auto cb = frame_callback;
-    if (cb) {
+#ifndef HAL_BOOTLOADER_BUILD
+    WITH_SEMAPHORE(callbacks.sem);
+#endif
+    bool added_to_rx_queue = false;
+    for (auto &cb : callbacks.cb) {
+        if (cb == nullptr) {
+            continue;
+        }
         if ((flags & IsMAVCAN) == 0) {
             cb(get_iface_num(), frame);
-        } else {
+        } else if (!added_to_rx_queue) {
             CanRxItem rx_item;
             rx_item.frame = frame;
             rx_item.timestamp_us = AP_HAL::micros64();
             rx_item.flags = AP_HAL::CANIface::IsMAVCAN;
             add_to_rx_queue(rx_item);
+            added_to_rx_queue = true;
         }
     }
     return 1;
 }
 
 /*
   register a callback for for sending CAN_FRAME messages
+  on success the returned callback_id can be used to unregister the callback
```

Comment:
```suggestion
  Register a callback for for sending CAN_FRAME messages.
  On success, the returned callback_id can be used to unregister the callback.
```
nit; These two comments bleed together

Reply:
changed